packages/myst-cli/src/process/file.ts

```diff
@@ -108,9 +110,84 @@ export function loadTexFile(
   return { kind: SourceFileKind.Article, mdast: tex.ast as GenericParent, frontmatter };
 }
 
-export function loadMySTJSON(content: string) {
-  const { mdast, frontmatter, kind } = JSON.parse(content);
-  return { kind: kind ?? SourceFileKind.Article, mdast, frontmatter };
+export function mystJSONValidationOpts(
+  vfile: VFile,
+  opts?: { property?: string; ruleId?: RuleId },
+): ValidationOptions {
+  return {
+    property: opts?.property ?? 'file',
+    file: vfile.path,
+    messages: {},
+    errorLogFn: (message: string) => {
+      fileError(vfile, message, { ruleId: opts?.ruleId ?? RuleId.mystJsonValid });
+    },
+    warningLogFn: (message: string) => {
+      fileWarn(vfile, message, { ruleId: opts?.ruleId ?? RuleId.mystJsonValid });
+    },
+  };
+}
+
+function validateMySTJSON(
+  input: any,
+  opts: ValidationOptions,
+): { mdast: GenericParent; kind: SourceFileKind; frontmatter: PageFrontmatter } {
+  const value = validateObjectKeys(
+    input,
+    { required: ['mdast'], optional: ['kind', 'frontmatter'] },
+    opts,
+  );
+  // Although this is required, we'll set it to an empty article by default
+  const mdast: GenericParent = value?.mdast ?? { type: 'root', children: [] };
+
+  let kind: undefined | SourceFileKind;
+  if (defined(input.kind)) {
+    kind = validateEnum<SourceFileKind>(input.kind, {
+      ...incrementOptions('kind', opts),
+      enum: SourceFileKind,
+    });
+  }
+  let frontmatter: undefined | PageFrontmatter;
+  if (defined(input.frontmatter)) {
+    frontmatter = validatePageFrontmatter(input.frontmatter, incrementOptions('frontmatter', opts));
+  }
+
+  return {
+    mdast,
+    kind: kind ?? SourceFileKind.Article,
+    frontmatter: frontmatter ?? {},
+  };
+}
+
+export function loadMySTJSON(
+  session: ISession,
+  content: string,
+  file: string,
+  opts?: LoadFileOptions,
+) {
+  const vfile = new VFile();
+  vfile.path = file;
+
+  const rawData = JSON.parse(content);
+  const {
+    mdast,
+    kind,
+    frontmatter: jsonFrontmatter,
+  } = validateMySTJSON(rawData, mystJSONValidationOpts(vfile));
+  const { frontmatter: astFrontmatter, identifiers } = getPageFrontmatter(
+    session,
+    mdast,
+    vfile,
+    opts?.preFrontmatter,
+    opts?.keepTitleNode,
+  );
+  const frontmatter = fillProjectFrontmatter(
+    astFrontmatter,
+    jsonFrontmatter,
+    frontmatterValidationOpts(vfile),
+  );
```

Comment:
I'm not totally sure how this should work, but to me, it feels like `jsonFrontmatter` is intended to be `PageFrontmatter`, rather than `ProjectFrontmatter`, and it should take priority, rather than be used as a fallback.

If that interpretation is correct, then I would use `jsonFrontmatter` as `preFrontmatter`, which would then potentially impact the `getPageFrontmatter` function... Something like:
```suggestion
  const { frontmatter, identifiers } = getPageFrontmatter(
    session,
    mdast,
    vfile,
    { ...astFrontmatter, ...opts?.preFrontmatter },
    opts?.keepTitleNode,
  );
```

Reply:
Ah, this was a good spot. I indeed wanted `jsonFrontmatter` to take precedence. I pulled in the `fill` function because I wanted to avoid a spread operator (thinking it was wrong to clobber things), but it looks like that's what we do internally anyway in `getPageFrontmatter`. So, I've made the changes!